### PR TITLE
Update `Marshal.QueryInterface()` argument modifier

### LINKED ISSUE #91981: .NET 8 RC1 can not compile with Windows app sdk and CsWinRT
### Description

.NET 8 RC1 can not compile with Windows app sdk and CsWinRT

In.net8 RC1, the Marshal.QueryInterface() method is defined from
public static int QueryInterface (IntPtr pUnk, ref Guid iid, out IntPtr ppv);
Become a
public static int QueryInterface (IntPtr pUnk, in Guid iid, out IntPtr ppv);
Please update the content in WinRT.cs immediately, otherwise the application will not compile

image

### Reproduction Steps

In describe bug

### Expected behavior

In describe bug

### Actual behavior

The Marshal.QueryInterface() method should be compatible with both ref and in arguments

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 8 RC1
Windows App SDK 1.4.0
CSWINRT 2.0.3

### Other information

None

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes #91981

Change `in` modifier to `ref readonly` to avoid warnings in existing interop code.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/IIUnknownStrategy.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ public unsafe interface IIUnknownStrategy
         /// <param name="iid">The IID (Interface ID) to query for.</param>
         /// <param name="ppObj">The resulting interface.</param>
         /// <returns>Returns an HRESULT represents the success of the operation.</returns>
-        /// <seealso cref="Marshal.QueryInterface(nint, in Guid, out nint)"/>
+        /// <seealso cref="Marshal.QueryInterface(nint, ref readonly Guid, out nint)"/>
         public int QueryInterface(void* instancePtr, in Guid iid, out void* ppObj);
```
Comment (stephentoub) [inline src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/IIUnknownStrategy.cs:36]: `in` is correct here?

**Review by stephentoub — APPROVED** [review]
cc: @tannergooding

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/IIUnknownStrategy.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/IIUnknownStrategy.cs:36]: That one is. It is a new API and we designed it to be `in`.

**Comment by jkotas** [comment]
> Change in modifier to ref readonly to avoid warnings in existing interop code.

I thought that it is "by design" that the change will produce warnings in existing code. There are other APIs with the same problem. For example, this compiled fine in .NET 7 and it will produce the exact same warning in .NET 8 as the Marshal.QueryInterface case.

```csharp
var span = new byte[10];
int value = 42;
MemoryMarshal.Write<int>(span, ref value);
```

If we are not happy with this change, should be roll it back in all APIs that were affected?

**Comment by AaronRobinsonMSFT** [comment]
> If we are not happy with this change, should be roll it back in all APIs that were affected?

I think in general I would agree that introducing new warnings should be avoided, but I do see the benefit in most cases. My specific concern here is the legacy nature of the `Marshal.QueryInterface()` API. I would be surprised if we see many users complain about the `MemoryMarshal.Write<int>()` APIs, but the `Marshal.QueryInterface()` API is old enough and in code that people rarely want to touch at all.

My calculus is on the nature of the `Marshal.QueryInterface()` API, it is very old and used in code that is rarely touched, and it adds no benefit to the interop scenario. Therefore any impact should be weight more.

**Comment by jkotas** [comment]
These two changes from https://github.com/dotnet/runtime/pull/89736#issue-1829841309 can introduce warnings in existing code:
- An API was changed from in to ref readonly, users will now be asked to use in explicitly
- An API was changed from ref to in, users will now be asked to use in or no longer specify ref

**Comment by jkotas** [comment]
> the legacy nature of the Marshal.QueryInterface() API

Ok, makes sense. 

It is unfortunate that we will end up with no consistency in how Guids are treated by various QI APIs:
- `ICustomQueryInterface.GetInterface` - ref Guid
- `Marshal.QueryInterface` - ref readonly Guid
- `IIUnknownStrategy.QueryInterface` - in Guid

**Comment by jkotas** [comment]
cc @Sergio0694

**Comment by Sergio0694** [comment]
Oh, well this is unfortunate 🥲

I was hoping starting with .NET 8 we'd be able to just pass `Type.GUID` without having to manually copy to a local first, which would've been nice.

**Comment by Sergio0694** [comment]
If people are _actually_ failing to build, isn't that simply because they're on an older version of Roslyn? The changes needed to properly support `ref readonly` didn't make it to VS 2022 Preview just yet I think. Like, the build error in #91981 will not happen if you're using an up to date version of Roslyn, it will only generate a warning (see language spec for `ref readonly`). Considering that, are we sure we actually do need to revert this signature change? The fact some new warnings would be generated was discussed and approved during API review, but there's no actual breaking changes if you have updated tooling 🤔

**Comment by hamarb123** [comment]
This should be fine to have as `in`. There should only be an issue if they've converted the warning into an error. We should ask them if they've turned that on (if they've got this on, then it's no different to other new warnings suddenly becoming errors on new TFMs) AND check if they're on an up-to-date compiler, as an older compiler would obviously have issues. And if that's not the case, ping the roslyn team to see if they know what's gone wrong, instead of merging this since it is supposed to be fine.

To be clear: this should not be merged on any version, because the issue shouldn't exist to start with. There is something else that has gone wrong. Either it's that they've enabled WarningsAsErrors (in which case they're being told about some unideal usage of an API as a warning, which they've opted to convert into an error), they've got an out-of-date compiler (either their fault, or the latest .NET 8 build has an old compiler which doesn't support the feature properly), or there's a bug in roslyn (which would obviously need to be fixed).

On a side note, this is why I suggested to the roslyn team on the issue (iirc I'm pretty sure I did anyway) that there should be no warning when using `ref` on an `in` parameter. If the user's issue is caused by WarningsAsErrors or similar, then would it be too late to fix the problem properly by changing this (i.e., by requesting the roslyn team to allow `ref` to be used for `in` parameters without a warning)?

**Comment by stephentoub** [comment]
> The fact some new warnings would be generated was discussed and approved during API review

In general, yes. We didn't go API by API and discuss every single one being changed. And for things in interop, the folks responsible for interop (aka Aaron) get to weigh in (and apparently hadn't been aware of this).

> there's no actual breaking changes if you have updated tooling

Whether or not it's something we consider to be "breaking", it's still friction. And the person responsible for our interop system is saying that friction isn't warranted.

> WarningsAsErrors / they've got an out-of-date compiler / or there's a bug in roslyn

Or `LangVersion` is set to something 11 or lower. It's a supported scenario to upgrade your TFM but not your LangVersion, and in that case, this will end up being an error, not a warning, regardless of WarningsAsErrors.

**Comment by Sergio0694** [comment]
> "Whether or not it's something we consider to be "breaking", it's still friction. And the person responsible for our interop system is saying that friction isn't warranted."

Sorry, let me rephrase that. Of course I do respect what Aaron is saying here, all I'm trying to say is just, given this PR is marked as a fix for an issue which is just a consequence of that user trying to compile with outdated tooling, and not a real break caused by the signature change when using C# 12, is it possible that the apparent need to revert this might've been influenced by that? As in, that considering the additional context (and the fact that user would not in fact see a build error as soon as they update Roslyn to latest), is there a chance undoing this might not actually be needed? 😅

> "It's a supported scenario to upgrade your TFM but not your LangVersion, and in that case, this will end up being an error, not a warning, regardless of WarningsAsErrors."

This is more of a side note, but I find this to be a bit confusing. One of the main arguments for including those relaxed combinations of ref modifiers as part of the spec was specifically to allow changing the signature of APIs like this without it being a source breaking change. If se say that upgrading TFM without keeping the language version in sync is a perfectly supported scenario, doesn't that kinda just invalidate all of that? Because you'd still end up in a situation where you'd have users on an "officially supported" scenario getting (unwanted) breaking changes when migrating to .NET 8 🤔

**Comment by stephentoub** [comment]
> doesn't that kinda just invalidate all of that?

No, it's just one more thing to factor in. It needs to be worth it. And Aaron is saying it's not worth it. There's a scale here and with all changes that introduce friction; this isn't something where everything is definitively 100% right or 100% wrong.

**Comment by stephentoub** [comment]
cc: @jaredpar

**Review by tannergooding — APPROVED** [review]
Seems good to me overall.

I'm not particularly a fan of avoiding new warnings, but I understand why there is a desire to do it here in particular.

It being `ref readonly` notably also doesn't block `typeof(T).GUID`, it just transfers the warning there instead which the people who want to explicit take advantage of this can suppress instead.

This means by changing from `ref` to `ref readonly` (rather than `in`) here on this very old API means we maintain the status quo for existing code, we improve the experience for new code by removing the need to `Unsafe.As` to strip the readonlyness, and we allow power users to still pass rvalues if they are willing to suppress a warning.
